pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+	pools := sets.NewString()
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Filter EIPs from nodeLabels failed")
+		return pools
+	}
+	for _, eip := range eips {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(eip.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) members() sets.String {
+	return sets.NewString(c.allClusterMembers()...)
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalIPPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+
+	n, err := c.mList.Join(c.members().List())
+	if err != nil {
+		klog.ErrorS(err, "Join cluster failed", "num", n)
+	}
+	if err := c.initConsistentHash(); err != nil {
+		klog.ErrorS(err, "Init consistentHashMap error")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.handleClusterNodeEvents(&nodeEvent)
+		}
+	}
+}
+
+func (c *Cluster) initConsistentHash() error {
+	node, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return err
+	}
+	eips := c.filterEIPsFromNodeLabels(node)
+	c.enqueueEIPs(eips)
+	return nil
+}
+
+func (c *Cluster) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Cluster) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (externalIPPool name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncConsistenHash(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing failed, requeue", "eip", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistenHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "eip", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	c.consistentHashRWMutex.Lock()
+	defer c.consistentHashRWMutex.Unlock()
```

Comment:
I think the critical section is larger that it needs to be and will actually cause all workers execute serially. It only needs to lock when it updates the map, i.e. before L352 and in `syncConsistentHashByEIP`. Note that the workqueue guarantees that a pool will only be processed by a single worker at any given time so you don't need to worry about the race condition between same item.

Reply:
got it, sounds good, will better as blow
```
	eip, err := c.externalIPPoolLister.Get(eipName)
	if err != nil {
		if errors.IsNotFound(err) {
			c.consistentHashRWMutex.Lock()
			defer c.consistentHashRWMutex.Unlock()
			delete(c.consistentHashMap, eipName)
			return nil
		}
		return err
	}

	// updateConsistentHash refresh the consistentHashMap
	updateConsistentHash := func(eip *v1alpha2.ExternalIPPool) error {
		nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
		if err != nil {
			return err
		}
		nodes, err := c.nodeLister.List(nodeSel)
		if err != nil {
			return err
		}
		aliveNodes := c.aliveNodes()
		// Node alive and Node labels matches with externalIPPool nodeSelector
		var aliveAndMatchedNodes []string
		for _, node := range nodes {
			nodeName := node.Name
			if aliveNodes.Has(nodeName) {
				aliveAndMatchedNodes = append(aliveAndMatchedNodes, nodeName)
			}
		}
		c.consistentHashRWMutex.Lock()
		defer c.consistentHashRWMutex.Unlock()
		c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
		c.consistentHashMap[eip.Name].Add(aliveAndMatchedNodes...)
		c.notify(eip.Name)
		return nil
	}
```